src/services/ContributorAuditService/providers/Bitbucket/BitbucketApiClient.ts

```diff
@@ -0,0 +1,248 @@
+        }
+        return Promise.reject(error);
+      },
+    );
+
+    return client;
+  }
+
+  async getBitbucketRepositories(): Promise<BitbucketRepository[]> {
+    const response = await this.client.get<BitBucketRepositoryApiResponse>(
+      `repositories/${this.workspace}`,
+    );
+
+    const repoResponse: BitBucketRepositoryApiResponse = response.data;
+
+    const repos: BitbucketRepository[] = repoResponse.values.filter(
+      (repo) => new Date(repo.updated_on) >= new Date(this.dateToFilter),
+    );
+
+    return repos;
+  }
+
+  async getBitbucketRepositoryContributors(
+    repository: BitbucketRepository,
+  ): Promise<IContributorAuditRepositories[]> {
+    const response = await this.client.get<BitbucketCommitsApiResponse>(
+      `repositories/${this.workspace}/${repository.name}/commits`,
+    );
+
+    response.data.values = response.data.values.filter(
+      (commit) => new Date(commit.date) >= new Date(this.dateToFilter),
+    );
+
+    const commits: BitbucketCommitsApiResponse = response.data;
+
+    const contributors = commits.values.reduce<IContributorAuditRepositories[]>((acc, commit) => {
+      const username = commit.author.user ? commit.author.user.display_name : "Unknown Author";
+      const commitDate = commit.date;
+
+      const repo: IContributorAuditRepository = {
+        id: repository.uuid,
+        name: repository.name,
+        lastCommit: commitDate,
+        isPrivate: repository.is_private,
+      };
+
+      let contributor = acc.find((c) => c.username === username);
+
+      if (!contributor) {
+        contributor = { username, repositories: [repo] };
+        acc.push(contributor);
+      } else {
+        const existingRepository = contributor.repositories.find((r) => r.id === repo.id);
+        if (!existingRepository) {
+          contributor.repositories.push(repo);
+        } else {
+          if (new Date(existingRepository.lastCommit) < new Date(commitDate)) {
+            existingRepository.lastCommit = commitDate;
+          }
+        }
+      }
+
+      return acc;
+    }, []);
+
+    return contributors;
+  }
+
+  private static async handlePagination<T extends BaseBitBucketApiResponse>(
+    response: AxiosResponse<T>,
+    client: AxiosInstance,
+    dateToFilter: string,
+  ): Promise<AxiosResponse<T>> {
+    let data = response.data;
+    let nextUrl = data.next;
+    let notOnDateRange = false;
+
+    while (nextUrl && !notOnDateRange) {
+      soosLogger.verboseDebug("Fetching next page", nextUrl);
+      const nextPageResponse = await client.get<T>(nextUrl);
+
+      if (data.type === BitbucketResponseTypes.Commit) {
+        data.values = (data.values as BitbucketCommit[]).concat(
+          nextPageResponse.data.values as BitbucketCommit[],
+        );
+        notOnDateRange = (data.values as BitbucketCommit[]).some(
+          (commit) => new Date(commit.date) < new Date(dateToFilter),
```

Comment:
for these date checks is that we have all of that type gymnastics that needs to be cleaned up currently. these endpoints don't support calling them with a parameter to limit the date and the fields to filter are called differently on each response type.

Reply:
the generics aren't worth it - there's only two calls - just repeat the logic and use helper functions if you need for date checking